### DISCUSSION TITLE
Suggestion: improve ergonomics of `*InfiniteOptions` types

### DISCUSSION THREAD

**Discussion by OliverJAsh** [TanStack/query#8484] [General]
### Describe the bug

[The docs](https://tanstack.com/query/latest/docs/framework/react/guides/query-options) suggest using `infiniteQueryOptions` to abstract infinite query options. For example:

```ts
import { infiniteQueryOptions } from '@tanstack/react-query';

export const queryOptions = infiniteQueryOptions({
  queryKey: ['foo'],
  queryFn: ({ pageParam }) => Promise.resolve({ foo: 'bar' }),
  initialPageParam: 1,
  getNextPageParam: (lastPage, allPages, lastPageParam) => lastPageParam + 1,
});

```

However, when using TypeScript's isolated declarations (`isolatedDeclarations`), all module exports must have type annotations. For this reason it would be preferable to use a type annotation rather than the `infiniteQueryOptions` type constructor.

The problem with this, however, is it's very verbose:

```ts
import type {
  DefaultError,
  InfiniteData,
  QueryKey,
  UnusedSkipTokenInfiniteOptions,
} from '@tanstack/react-query';

type MyResponse = {
  foo: string;
};

export const queryOptions: UnusedSkipTokenInfiniteOptions<
  MyResponse,
  DefaultError,
  InfiniteData<MyResponse, number>,
  QueryKey,
  number
> = {
  queryKey: ['foo'],
  queryFn: ({ pageParam }) => Promise.resolve({ foo: 'bar' }),
  initialPageParam: 1,
  getNextPageParam: (lastPage, allPages, lastPageParam) => lastPageParam + 1,
};

```

Except for the first and last type parameters provided to `UnusedSkipTokenInfiniteOptions`, all the others could theoretically just use the defaults. However, we're forced to provide them because we need to provide `TPageParam` (the last type parameter), otherwise it would default to `unknown`.

Given more TypeScript users will likely be looking to enable isolated declarations in the future and thus be required to add type annotations, I wonder if we could make these types more ergonomic.

One idea would be to move the position of the `TPageParam` type parameter in the `*InfiniteOptions` types, so we can benefit from defaults. For example, if it was moved from last to second position then the example above would look much better:

```ts
import type {
  DefaultError,
  InfiniteData,
  QueryKey,
  UnusedSkipTokenInfiniteOptions,
} from '@tanstack/react-query';

type MyResponse = {
  foo: string;
};

export const queryOptions: UnusedSkipTokenInfiniteOptions<
  MyResponse,
  number
> = {
  queryKey: ['foo'],
  queryFn: ({ pageParam }) => Promise.resolve({ foo: 'bar' }),
  initialPageParam: 1,
  getNextPageParam: (lastPage, allPages, lastPageParam) => lastPageParam + 1,
};

```

```diff
-type UnusedSkipTokenInfiniteOptions<TQueryFnData, TError = DefaultError, TData = InfiniteData<TQueryFnData>, TQueryKey extends QueryKey = QueryKey, TPageParam = unknown> = OmitKeyof<UseInfiniteQueryOptions<TQueryFnData, TError, TData, TQueryFnData, TQueryKey, TPageParam>, 'queryFn'> & {
    queryFn?: Exclude<UseInfiniteQueryOptions<TQueryFnData, TError, TData, TQueryFnData, TQueryKey, TPageParam>['queryFn'], SkipToken | undefined>;
};
+type UnusedSkipTokenInfiniteOptions<TQueryFnData, TPageParam = unknown, TError = DefaultError, TData = InfiniteData<TQueryFnData, TPageParam>, TQueryKey extends QueryKey = QueryKey> = OmitKeyof<UseInfiniteQueryOptions<TQueryFnData, TError, TData, TQueryFnData, TQueryKey, TPageParam>, 'queryFn'> & {
    queryFn?: Exclude<UseInfiniteQueryOptions<TQueryFnData, TError, TData, TQueryFnData, TQueryKey, TPageParam>['queryFn'], SkipToken | undefined>;
};
```

### Your minimal, reproducible example

N/A

### Steps to reproduce

N/A

### Expected behavior

N/A

### How often does this bug happen?

None

### Platform

N/A

### Tanstack Query adapter

None

### TanStack Query version

N/A

**Comment by TkDodo** [comment]
using the defaults doesn’t really help much because using them will make features where type inference is necessary, like `select`, not work.

We use `isolatedModules` without explicitly giving everything a return type annotation, so I’m not really sure why inference can’t be used. We also can’t change the type param order because that would be a breaking change.

Moving to a discussion.

**Reply by OliverJAsh** [reply]
Sorry, I was talking about `isolatedDeclarations`, not `isolatedModules`. Hopefully that clarifies things?

**Reply by samhh** [reply]
Such inference is incompatible with isolated declarations: https://www.typescriptlang.org/docs/handbook/release-notes/typescript-5-5.html#isolated-declarations

There is already some early-stage tooling such as oxc-transform that leverages this feature, dramatically speeding up typechecking in large projects by essentially eliminating the time taken to perform declaration emit.

This is still quite immature in the ecosystem but it's the direction of travel.

**Comment by karlshea** [comment]
Hijacking this a bit because I'm having a similar problem but even passing the types isn't helping. I'm trying to convert a component to TypeScript and I cannot get type generics to work with `useInfiniteQuery`. I've trimmed this down to as minimal of an example as I can:

```typescript
import React, { ReactNode } from 'react';
import {
  DefaultError,
  InfiniteData,
  infiniteQueryOptions,
  QueryKey,
  useInfiniteQuery,
} from '@tanstack/react-query';

type TestPageParams = {};

export interface InfiniteScrollProps<T, P extends TestPageParams> {
  children: (data: InfiniteData<T, P> | undefined) => ReactNode;
  queryFn: (pageParam: P) => Promise<T>;
  queryKey: QueryKey;
  initialPageParam: P;
}

export const InfiniteScroll = <T, P extends TestPageParams>({
  children,
  queryFn,
  queryKey,
  initialPageParam,
}: InfiniteScrollProps<T, P>) => {
  const { data } = useInfiniteQuery(
    infiniteQueryOptions({
      queryKey: [...queryKey, initialPageParam],
      queryFn: ({ pageParam }) => queryFn(pageParam),
      initialPageParam,
      getNextPageParam: (lastPage, allPages, lastPageParam) => lastPageParam,
    })
  );

  return <>{children(data)}</>;
};
```

I'm getting "TS2345: Argument of type unknown is not assignable to parameter of type P. P could be instantiated with an arbitrary type which could be unrelated to unknown" on the `pageParam` argument to `queryFn`, and hovering over `useInfiniteQuery` is giving me this inferred type signature:

```
export function useInfiniteQuery<T, Error, InfiniteData<T, unknown>, unknown[], P>(
  options: UndefinedInitialDataInfiniteOptions<T, Error, InfiniteData<T, unknown>, unknown[], P>, 
  queryClient?: QueryClient,
): UseInfiniteQueryResult<...>
```

Changing the code to `useInfiniteQuery<T, DefaultError, InfiniteData<T, P>, QueryKey, P>` is giving me an absolutely massive error about what `infiniteQueryOptions` is returning (no overload matches this call) about QueryKey. Changing that back to `unknown[]` just gets me right back to the original `pageParam` unknown problem. 

What am I missing here? It feels like `TPageParam` isn't getting passed all the way through somewhere.

Edit: tried using OP's method and having the same issue:

```typescript
const queryOptions: UnusedSkipTokenInfiniteOptions<T, P> = {
  queryKey: queryKey,
  queryFn: ({ pageParam }) => queryFn(pageParam),
  initialPageParam,
  getNextPageParam: (lastPage, allPages, lastPageParam) => lastPageParam,
};
```

**Reply by TkDodo** [reply]
this is an interesting case, because I think there is something wrong with our `QueryFunctionContext` type.

Here’s a [TS playground](https://www.typescriptlang.org/play/?#code/JYWwDg9gTgLgBAJQKYEMDGMA0cDejUYByEAJknAL5wBmUEIcA5FATIwNwBQoksuncQXAAiSaigCuAGxgBRKHSiYBQgJIA7asHXAYSYShgplQuNq069ARQlIoATwDyYGMAjqAziaE279gNJI9t6CEh5IGha6SL4OylS09EwAAkaeRmgA1gD0LOgwALQAjrYOHJycMPZg5AAqSB4wAAooAOZILVAoIB5wALy4FFycSAAevPDaelDiaOSR2tEAymh0UlJNdGAeADy12E1wY3rqJL31jS3tnd0eAHz8pmgAFsBSJCzqAFxwABQkhhQPwWln0gL2BweAB84BJTmJtEgSABKfoPZD5YhkLimEp+ABi3z+YDaHRQXRAPyaqL6D029GA4T2dxxQjxDkC9h+sQCQVZgkWrhQG1JN0pcCaXAoFTGEzgaHcjTgIOWqwg636cAhEqOoxOZzgF2aovJtzuvxwKkELzeHyQ6hCcHZ9kJjudnMdguAwquZIp8WBmkFSBWaw2W12+wldxpD0tT0V8DwAKMlE1YQiQdBPN+VtMj3zhfdQR+AG0AHSV4vBMyWb0i66mkAAXUdhcEzsJP1+CvUej1sadpRd6h77n7MHLJMbFORbfbXp9Jv9efbcHaMEIx19Yu7UhQl1J2GFDYa2H3h5n3UHF+NV5A88E0tMyOGphYMAkUHUWruOBt7yfP8gLIhQOzZCynBDEAA).

When we look at how `queryFn` is inferred, we can see it’s:

```
(property) queryFn?: unique symbol | QueryFunction<T, unknown[], P> | undefined
```

here, `P` is still there as a type for `pageParam`. Even when we look at the context that gets passed, it’s  still there:

```
(parameter) context: QueryFunctionContext<unknown[], P>
```

However, when we look at the `context.param` then is:

```
(property) pageParam?: unknown
```

if we look at `context.direction`, it’s also unknown:

```
(property) direction?: unknown
```

This makes me believe our `QueryFunctionContext` type has a problem:

https://github.com/TanStack/query/blob/1e73d3a3ad245d2b17dd1d72b01d3841b11d3a01/packages/query-core/src/types.ts#L104-L129

It seems to go into the `[TPageParam] extends [never]` branch, which I don’t think it should, because `TPageParam` is clearly of type `P`, not `never` :/

Sorry for the ping @Andarist but you usually know pretty quickly why that is 🙏

**Reply by Andarist** [reply]
`TPageParam` could be `never` though. It's not that it goes into the first branch - it goes to... both. So the computed type there is `P | unknown` but that just reduces to `unknown`.

Simplified repro of the problem above: [TS playground](https://www.typescriptlang.org/play/?ts=5.7.3#code/C4TwDgpgBAigrhATiAYnAdgY2ASwPboDCBwEAHsADwAqACgIYDmEDi9AtlALxToQBuSAHzcoAbTpMW9NuwC6UcqXQATAM7i+gxHIBQUKAH4oAbyhgprDoYBcUDAGt0eAO7ooAX31Q7Zi8yt2O0kAmQ5PAG5dXVBIWARkNCxcAhoGUNlRLWFRAApvTBIlO3gkVAxsfCIiijTLMPYhABpdAEpuEX48HBUo3RUITAAbGWgAMwqU9zg1CABJdDGcdBxSUuQ6jPCebMQhXIB9X29l1Zx6IfTpWWCrwKiDAEcE1HQSl6TK1JDrjiEojytOxdHp9chgPCIYBQCbJKpQUhqKi0RQUCCqDQmDz7EwnFa4C53BotDzHAynAmXeo3KC0AHtXEGGbzRYUiDrEC5RkGKAU85UrbsFo8qDPMooN5QXKFdCkCjtLgibkimVI0zmanhDzcbwigwAen1eoAeoZdXqZXLgBEoIaoLNoMAABaO8DQPBjBEuqAAA0tSh9UBdiAgTV5wAA5Bp6PYVgRzQajabzR5hZ5WgDogNhqMoKroc5nctGHZdn1EVRdji8WdCZqgrw8EX0IwSRndEA)

**Reply by TkDodo** [reply]
hmm, so could we define a `P` generic that is _not_ never 😅 ? The current example does:

```
type TestPageParams = {};

export const InfiniteScroll = <T, P extends TestPageParams>({
```

so could we say that `T extends TestPageParams` but not `never` ?

Tbh, we only made the conditional type internally to differentiate between `useQuery` and `useInfiniteQuery`. They both have a `queryFn`, but with a different context that gets passed. Maybe there’s a better way for that, too?

**Reply by Andarist** [reply]
> hmm, so could we define a P generic that is not never 😅 ? 

Not really. There are no negated types in TS and you can only mimic them with conditional types. But those are deferred at generic level (like here) so you end up with issues like this.

> Tbh, we only made the conditional type internally to differentiate between useQuery and useInfiniteQuery. They both have a queryFn, but with a different context that gets passed. Maybe there’s a better way for that, too?

I'd have to know more about your use case. Could you wrap it in a self-isolated TS playground that would not depend on React Query types?

**Reply by karlshea** [reply]
Is distributivity not desired for this type? I'm a little hazy on it, but in the playground from @Andarist if I change the type to this (removing the square brackets) I do get the right type on `pageParam`:

```
type QueryFunctionContext<TPageParam = never> = TPageParam extends never
  ? { pageParam?: unknown }
  : { pageParam: TPageParam };
```

**Reply by Andarist** [reply]
```ts
type Test<T> = T extends never ? 1 : 0
type Result = Test<never>
//   ^? type Result = never
```

A distributive conditional type is applied separately to all union constituents. Then the results are put "back" into a union. `never` is an empty union (in a sense) so conditional type applied to never... just gives u `never` back. So yeah, a distributive conditional can't be used here while preserving the original intent

**Reply by karlshea** [reply]
So it would be breaking something somewhere else, just not this thing in particular?

**Reply by karlshea** [reply]
And then I guess do you have any suggestions on how I can solve this?

**Reply by TkDodo** [reply]
> I'd have to know more about your use case. Could you wrap it in a self-isolated TS playground that would not depend on React Query types?

I extended your simplified playground a bit to show both `useInfiniteQuery` and `useQuery`:

[TS Playground](https://www.typescriptlang.org/play/?ts=5.7.3&ssl=26&ssc=22&pln=25&pc=21#code/C4TwDgpgBAigrhATiAYnAdgY2ASwPboDCBwEAHsADwAq8SIA0hCADRTUAKAhgOYTeIuAWygBeKOggA3JAD4xUANqde-LoKEBdKOVLoAJgGclkmYk0AoKFAD8UAN5QAjgmRMQALnZ03zNmFUBYRsvUyQ2fRxECGx8dBCJaSQoAF8rKC9HF3p3L1pXRj8oAL4goTzuUvVhCKiY3AIvAHIAMzxEAHd1fSaoAB8oJoAjLkwAay7EHtSAbgsLUEhYArQsBvQaH0LWdkq1DQUwxHlxAAp0zBJdLy3V2IJidFIKTYL3NhUqjVkWCwBKMTyKR4HD6OYWfQxAA26mgLQw93QUDghggAEl0C0cOgcKQtq8ckVPvthIcksdTgB9TLpbK+TzeN5+dLY3E4LhQvZlCqBapCObWOmodA3FYI9YE+kfLl82RzFJ-LzA0HgyGYGHRKDwtZxZGo-H5QkgWRUmmCpkMw1S2krEXLeh3CVW7ayCwK+bkMDtYBa8W60iGKgcHQUCAGYz2FIm+wsnG4DkyjS-FJmqCs+Oc3kaLwceUAmPWFHozHpiBbU4F6xpuPszNfGrpc1GryKJrYrFxiBNbRcYyXdCB35V5y2rynfvPYAA0TySvD-uBhzFLOklJiRvDqAAei3m4AejYN-OrhQZtvd6joMAABZX8DQPAtKA36AAAwnulfUFv0TYuKaxhcMiOIEEe55QAeG4pEOqR-PK8xqhq0ALj66B4De2I8KE5LggGVBHNGsZsgmK7lBI6HXphyZwfMKEjvQABKECGHAUI+uIRblpWQq5EoTRCt2UC9lAKEwU2yAoHa44nlOgIOGBdFZBaqTrpuVY7vuh5qdYH6nuBl7Prez73lAj6GW+unAF+P4QH+wAAUJwFxGBGmQcO0Fun8QA)

Basically, we have two function: `useQuery` and `useInfiniteQuery`. Both have a `queryKey` and a `queryFn`, but only the infinite version does some more things with pageParams and the `TPageParam` type parameter. But they both have a `queryFn` that receives a context - that context has different values at runtime.

There is nothing to discriminate though, so we tried to infer the correct type of the context by checking the `TPageParam` type parameter with `[TPageParam] extends [never]`. You can see that `useQuery` doesn’t pass the `TPageParam` type parameter to the `QueryFunction` type, so the `never` default is used.

**Reply by Andarist** [reply]
I might be missing some context on this but this particular example works if you stick to type composition/inheritance instead of "routing" in your types: [TS playground](https://www.typescriptlang.org/play/?ts=5.7.3#code/C4TwDgpgBAigrhATiAYnAdgY2ASwPboDCBwEAHsADwAq8SIA0hCAHxQC8UA3lAI4LImIAFxRaAxsygBfANwAoeaEiwJaLLgI06g5m04AKeVCiYS5YKJ2oM2fEXMVtEoSwA08gJQc2ANzw4ACYKSuDQAJLoAGY46Dik1up2BMTopE7i9EJuYgAKAIYA5hAFiPkAtvrcxnwuzKKZuiAKJmBFJfll5Q0FxaUVLVCBOIgQyeiiAERReIgA7p2Bk1AAPlCTAEb5mADWC4hLCnKKyhHRsfEQibaa6M5ZzDnUvR1dVUYmZmkWopExcQk1Dd7Kl0lRGpIQE8Xv1Kh5vOw-AFgopAmMADadaBRYEEKBwADOED+F0B9HuTWh7VhLAMAH1RFwavwHiIxNYhINSTh8uiYZ0Kj1qQLyoMWcgUBMoCSAVcgRp7BTIVS+iKWEdPKJ-EEQmjMJjRlAcQq8YS5eSIa56YzmXU2ZbmGK1FLria7g7WBqtciQuQwLNgEbcegoKQCVRclALBB0IECdxpLSmSZubz+V0PNIbSmATy+cKuqJchrqiYzTLLtYDMmTFBU-nVRmaiZxZDRABtSaxf6XSYAXSg+XjX3DHlrtXoktEBi+YIRbBr44A9EvxyYAHoAfmbtZHgZ4bUbFRkHFMjmAg2Xq7XW530jHMk8R1RGKxZ-Q4ag6DwwAAFrFClEdAIF8JAQjDKhgNAxAkxqet00FL8f3-dBCkzJ9FD3CdkAAJQgAk4HRQNODNKsa1bIQO0mVt+0HYcCFHW1JylGdz3nUtaxXNcoFvccsJ4CipGkU9ZwsS8oC4m9txMe95GkDCgA)

> There is nothing to discriminate though

Couldn't you discriminate based on `initialPageParam` in the parent object?

**Reply by TkDodo** [reply]
I think we could, but we currently don’t have a different type for `queryFunction` on Infinite Queries. I think overall, this would likely be the better approach.

Alternatively, could we just use a `unique symbol` type instead of `never` to check against? I think it would just be a matter of setting the default value for the type `TPageParam`. I was hoping this would work but it doesn’t seem so: [TS playground](https://www.typescriptlang.org/play/?ts=5.7.3#code/C4TwDgpgBAigrhATiAYnAdgY2ASwPboDCBwEAHsADwAq8SIA0hCADRTUAKAhgOYTeIuAWygBeKADk83PgOEA+MVADanXvy6ChAXSjlS6ACYBnFVJkat2gFBQoAfigBvKAEcEyJiABc7Op+Y2MHU5IXtfdAgANyQ2QxxECGx8dHCoSJjEKABfWyhfF3d6L19aD0ZAqGDZTWFSi1C4hKTcAl8AcgAzPEQAd01DdqgAHyh2gCMuTABrfsRBnIBua2tMAmNgdOkQ2pFxAGUQIXG8ABsACgBKa1BISW2arSVbiDxOrYbdlZfYcrQsVroGj+CqsdifJ7icw7LSKcTnPJrdCkCi+EH-ZIEYjI-TA8peNhqR4KFjWS5iRRRPA4QzLayGJKnTTQToYTHoKBwYwQACS6E6OHQOFIILxxUqRMswiU0OJQnk5wA+gU8kUAj4-PjAnlBcKcFxThC6uCYcJlnY1ah0Gi-mzAWL1YSjfLltlLr4qTS6QzMEzElBWQCUpzuaKyuKQArlc5VVqNeHHbH6Chrb9k3aUg7QfJrG6PdTDCtyGAeptA+yoKQNpQOAqnDqhbgDc7SdkVXZdU3DaahL4OK7yfW7FzefzOxAQech3YoJ39d25aSZ24475lO1BQLGxB2rouKYkRslzPLSnfOckSjgOTRIpp8vD5sXNUpSJsmI8su7AB6b9fgB69ifl+l76IsUC-lA3LQMAAAWMHgNAbyVvBUAAAagRQaFQPBiRsMK7SmFwnJCgQwE-n+gHAdkx5uq63wQBsU4NnqzY9h0+6GJ07StpcdKPiu9AAEqMXApybOII6TtOlolCo7SWruUD7lAj7HhafypheJD6Ded7kap6xPoJ6o5B+X4zpBy5URZdiYcA4GQdBKEIXcyFwdAGE6VhOFIBA+HAIRWwkSkBmQTZM40bmlxAA)

**Reply by Andarist** [reply]
Ultimately it doesn't what you check against. The only thing that matters here is the conditional type presence.

**Reply by karlshea** [reply]
Just an update, this is what got it working for us. We had to cast `TPageParam` and not use the `infiniteQueryOptions()` helper.

```tsx
const queryOptions: UnusedSkipTokenInfiniteOptions<
    ODataListResponse<T>,
    DefaultError,
    InfiniteData<ODataListResponse<T>, TPageParam>,
    QueryKey,
    TPageParam
  > = {
    queryKey: [...queryKey, initialPageParam],
    queryFn: (context) => queryFn(context.pageParam as TPageParam),
    initialPageParam,
    getNextPageParam: (lastPage, allPages, lastPageParam) => { ... }
    enabled,
  };

  const {
    isLoading,
    isRefetching,
    data,
    isFetchingNextPage,
    fetchNextPage,
    hasNextPage,
    error,
  } = useInfiniteQuery(queryOptions);
```
